core/base/src/TDirectory.cxx

```diff
@@ -272,9 +283,25 @@ void TDirectory::CleanTargets()
          fContext = next;
       }
 
-      for(auto ptr : fGDirectories) {
-         if (ptr->load() == this) {
-            (*ptr) = nullptr;
+      for (auto &ptr : fGDirectories) {
```

Comment:
I don't understand this part. This is *very* complex, and the amount of comments that explain intent and design of this seems inadequate for people like me trying to understand this.